### PR TITLE
Patterns Observer

### PR DESCRIPTION
**Pull request by AntonTmsh**
4.1. Observer

### CODE REVIEW

File: Observer/Observer/Observer/Program.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Observer
+{
+    class Program
+    {
+        static void Main(string[] args)
+        {
```
Comment (fojman) [inline Observer/Observer/Observer/Program.cs:8]: why main empty?

File: Observer/Observer/ConsoleObserver/Program.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    {
+        static void Main(string[] args)
+        {
+            ConsoleReader conReader = new ConsoleReader();
+            EventObserver eventObserver = new EventObserver(conReader);
+            Console.WriteLine("Inpute quit to stop");
+            do
+            {
+                conReader.InputString = Console.ReadLine();
+            } while (conReader.Flag);
```
Comment (fojman) [inline Observer/Observer/ConsoleObserver/Program.cs]: Sorry but didnt get what this code does? i guess it's just waiting for user input and that's it?

**Review by fojman — CHANGES_REQUESTED** [review]

File: Observer/Observer/ConsoleObserver/Program.cs (hunk 2 above)
Comment (AntonTmsh) [reply Observer/Observer/ConsoleObserver/Program.cs]: Write a class which listens to console and raise an event if 'quit' is written in console

File: Observer/Observer/Observer/Program.cs (hunk 1 above)
Comment (AntonTmsh) [reply Observer/Observer/Observer/Program.cs:8]: fixed

File: Observer/Observer/ConsoleObserver/ConsoleReader.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    {
+        private string inputString;
+        public bool Flag { get; set; }
+
+        public string InputString
+        {
+            get { return inputString; }
+            set
+            {
+                inputString = value;
```
Comment (fojman) [inline Observer/Observer/ConsoleObserver/ConsoleReader.cs]: 4.1.1. Write a class which listens to console and raise an event if 'quit' is written in console:
- we need to raise an event if 'quit' word put in and you raise it for any input, am i right?

File: Observer/Observer/ConsoleObserver/AbstractObserver.cs [hunk 4]
```diff
@@ -0,0 +1,13 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Net.Mentoring.Patterns.ConsoleObserver
+{
+    public abstract class  AbstractObserver
```
Comment (fojman) [inline Observer/Observer/ConsoleObserver/AbstractObserver.cs]: why do we need an abstract class here? it looks like `interface`

File: Observer/Observer/ConsoleObserver/AbstractSubject.cs [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Net.Mentoring.Patterns.ConsoleObserver
+{
+    public abstract class AbstractSubject
+    {
+        List<AbstractObserver> observers = new List<AbstractObserver>();
```
Comment (fojman) [inline Observer/Observer/ConsoleObserver/AbstractSubject.cs]: better to use access midifer explicitly

File: Observer/Observer/InterfacesObserver/GoogleMonitor.cs [hunk 6]
```diff
@@ -0,0 +1,26 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace Net.Mentoring.Patterns.IObserver
```
Comment (fojman) [inline Observer/Observer/InterfacesObserver/GoogleMonitor.cs]: weird namespace name

File: Observer/Observer/Observer/SampleData.cs [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+{
+    class SampleData
+    {
+        private static decimal[] samplePrices = new decimal[] { 10.00m, 10.25m, 555.55m, 9.50m, 9.03m, 500.00m, 499.99m, 10.10m };
+        private static string[] sampleStocks = new string[] { "MSFT", "MSFT", "GOOG", "MSFT", "MSFT", "GOOG", "GOOG", "MSFT" };
+        public static IEnumerable<Stock> getNext()
+        {
+            for (int i = 0; i < samplePrices.Length; i++)
+            {
+                Stock s = new Stock();
```
Comment (fojman) [inline Observer/Observer/Observer/SampleData.cs]: use initialization list - i.e. new Stock{a  = b }

File: Observer/Observer/ConsoleObserver/Program.cs [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Net.Mentoring.Patterns.ConsoleObserver
+{
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            ConsoleReader conReader = new ConsoleReader();
```
Comment (fojman) [inline Observer/Observer/ConsoleObserver/Program.cs]: image
- Please fix spelling typos
- If you take into account input case - please tell the user this(or ignore case).

**Review by fojman — COMMENTED** [review]
Just to don't look at KB page every time.
Practice
4.1.1  Write a class which listens to console and raise an event if 'quit' is written in console
4.1.2  StockExchange on Delegates (it’s a wrong way, you expose delegate outside and external code can crash your internal state
4.1.3  StockExchange on Events
4.1.4  StockExchange on Interfaces

File: Observer/Observer/ConsoleObserver/ConsoleReader.cs (hunk 3 above)
Comment (AntonTmsh) [reply Observer/Observer/ConsoleObserver/ConsoleReader.cs]: I don't think that important from a study perspective, but you are right. Fixed.

File: Observer/Observer/ConsoleObserver/AbstractObserver.cs (hunk 4 above)
Comment (AntonTmsh) [reply Observer/Observer/ConsoleObserver/AbstractObserver.cs]: to show, a classic pattern model, could be an interface, fixed

File: Observer/Observer/ConsoleObserver/AbstractSubject.cs (hunk 5 above)
Comment (AntonTmsh) [reply Observer/Observer/ConsoleObserver/AbstractSubject.cs]: ok

File: Observer/Observer/InterfacesObserver/GoogleMonitor.cs (hunk 6 above)
Comment (AntonTmsh) [reply Observer/Observer/InterfacesObserver/GoogleMonitor.cs]: ok

File: Observer/Observer/Observer/SampleData.cs (hunk 7 above)
Comment (AntonTmsh) [reply Observer/Observer/Observer/SampleData.cs]: ok

File: Observer/Observer/ConsoleObserver/Program.cs (hunk 8 above)
Comment (AntonTmsh) [reply Observer/Observer/ConsoleObserver/Program.cs]: ok

**Pull request merged**